examples/bigquery_user_info_updater_tool/cron/bigquery_user_info_updater/tests/test_nested_user_info_updater.py

```diff
@@ -0,0 +1,255 @@
+
+         Returns:
+             True if test passes, else False.
+                 """
+        # Load a set of user updates to nested user_info_updates table.
+
+        self.load_json_to_bq(
+            filename='test_data/nested_data/user_info_updates_data.json',
+            table=self.dataset_ref.table(self.user_info_updates_id)
+        )
+
+        # Load data into the temp table and final table to simulate a previous
+        #  run.
+        self.load_json_to_bq(
+            filename='test_data/nested_data/'
+                     'temp_user_info_updates_initial.json',
+            table=self.dataset_ref.table(self.temp_user_info_updates_id)
+        )
+        self.load_json_to_bq(
+            filename='test_data/nested_data/user_info_final_initial.json',
+            table=self.dataset_ref.table(self.user_info_final_id)
+        )
+
+        # Run the UserInfoUpdater on the second set of updates.
+        test_updater = user_info_updater.UserInfoUpdater(
+            project_id,
+            self.dataset_id,
+            self.user_info_updates_id,
+            self.temp_user_info_updates_id,
+            self.user_info_final_id
+        )
+        update_query_creator = query_creator.QueryCreator(
+            schema_path=self.schema_path,
+            user_id_field_name='userId',
+            ingest_timestamp_field_name='ingestTimestamp',
+            project_id=project_id,
+            dataset_id=self.dataset_id,
+            updates_table_id=self.user_info_updates_id,
+            temp_updates_table_id=self.temp_user_info_updates_id,
+            final_table_id=self.user_info_final_id
+        )
+
+        gather_updates_query = update_query_creator.create_gather_updates_query()
+        test_updater.gather_updates(gather_updates_query)
+        merge_udpates_query = update_query_creator.create_merge_query()
+        test_updater.merge_updates(merge_udpates_query)
+
+        # Query the temp table to test that the gather_updates() function worked
+        temp_table_query_config = bigquery.QueryJobConfig()
+        temp_table_query_config.use_legacy_sql = False
+        temp_table_query = self.bq_client.query(
+            query='SELECT * FROM `{0:s}.{1:s}.{2:s}`'.format(
+                project_id,
+                self.dataset_id,
+                self.temp_user_info_updates_id
+            ),
+            job_config=temp_table_query_config,
+            location='US'
+        )
+        temp_table_query.result()
+        temp_table_results_df = temp_table_query.to_dataframe() \
+            .sort_values(by=['userId']).reset_index(drop=True)
+
+        # Gather expected results for comparison
+        abs_path = os.path.abspath(os.path.dirname(__file__))
+        expected_temp_data_file = os.path.join(
+            abs_path,
+            'test_data/nested_data/temp_user_info_updates_expected.json'
+        )
+        expected_temp_table_df = pd.read_json(
```

Comment:
jaketf on Aug 27, 2019 Member
you're repeating this "read a json to a dict" a lot. Just define a function
You're leaving files open.
use with context and json.load to load file like object (instead of json.loads to load a string)

let's DRY this up and define a convenience function.

def read_json(filename):
    """Deserialize a json file to dict.
    """
        with open(filename) as json_file:
            return json.load(json_file)